File: engine/argot/scoring/scorers/except_return_raise_ratio.py
"""Except-block return/raise ratio primitive.

Scalar: ratio of ``return_statement`` nodes to
``(return_statement + raise_statement | throw_statement)`` nodes that
appear inside ``except_clause`` (Python) or ``catch_clause`` (TypeScript)
handler subtrees.  Per-cluster baseline is ``(mean, std)`` across cluster
files.  Score-time contribution is a tail-z penalty clipped at
``cluster_bonus_clip``.

Ramp curve:
  contribution = min(cluster_bonus_clip, max(0.0, abs(tail_z) - 1.0))
  Ramps linearly from 0 at |z|=1σ to cluster_bonus_clip at
  (1 + cluster_bonus_clip)σ.  Two-sided: catches anomalously HIGH or
  LOW ratios.  Monotone and bounded.
"""

from __future__ import annotations

import statistics
from collections.abc import Iterable, Iterator
from dataclasses import dataclass
from pathlib import Path

from tree_sitter import Node

from argot.scoring.filters.typicality import _PY_PARSER, _TS_PARSER
from argot.scoring.scorers.shape_primitive import Language

# ---------------------------------------------------------------------------
# Baseline payload
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class _ExceptRatioBaseline:
    mean: float
    std: float


# ---------------------------------------------------------------------------
# AST helpers
# ---------------------------------------------------------------------------

_PY_HANDLER_TYPE = "except_clause"
_TS_HANDLER_TYPE = "catch_clause"
_RETURN_TYPE = "return_statement"
_PY_RAISE_TYPE = "raise_statement"
_TS_RAISE_TYPE = "throw_statement"


def _walk(root: Node) -> Iterator[Node]:
    """Pre-order DFS over the subtree rooted at *root*."""
    stack: list[Node] = [root]
    while stack:
        node = stack.pop()
        yield node
        stack.extend(reversed(node.children))


def _count_in_handlers(
    root: Node,
    handler_type: str,
    raise_type: str,
) -> tuple[int, int]:
    """Return ``(returns, raises)`` inside all handler blocks under *root*.

    Descends the full tree looking for ``handler_type`` nodes (except_clause
    / catch_clause), then counts return and raise nodes inside each one.
    Nested handlers are walked independently.
    """
    returns = 0
    raises = 0
    for node in _walk(root):
        if node.type == handler_type:
            for child in _walk(node):
                if child.type == _RETURN_TYPE:
                    returns += 1
                elif child.type == raise_type:
                    raises += 1
    return returns, raises


def _ratio_for_source(source: str, language: Language) -> float | None:
    """Compute except-block ratio for a single source string, or None to skip.

    Returns None when the source has no return/raise inside handler blocks
    (ratio is mathematically undefined — the hunk/file is not exercising
    the exception-handling pattern this primitive measures).
    """
    parser = _PY_PARSER if language == "python" else _TS_PARSER
    handler_type = _PY_HANDLER_TYPE if language == "python" else _TS_HANDLER_TYPE
    raise_type = _PY_RAISE_TYPE if language == "python" else _TS_RAISE_TYPE
    try:
        tree = parser.parse(source.encode("utf-8"))
    except Exception:  # noqa: BLE001
        return None
    returns, raises = _count_in_handlers(tree.root_node, handler_type, raise_type)
    total = returns + raises
    if total == 0:
        return None
    return returns / total


def _ratio_for_hunk(hunk_content: str) -> float | None:
    """Try Python grammar then TypeScript grammar; return ratio or None.

    The grammars are disjoint on handler-block node types (except_clause
    vs catch_clause), so probing both is safe: whichever grammar matches
    the hunk's syntax will find the relevant nodes; the other returns None.
    """
    ratio = _ratio_for_source(hunk_content, "python")
    if ratio is not None:
        return ratio
    return _ratio_for_source(hunk_content, "typescript")


# Minimum number of cluster files that must have a defined ratio before
# the cluster baseline is trusted.
_MIN_VALID_FILES = 3


# ---------------------------------------------------------------------------
# Primitive
# ---------------------------------------------------------------------------


@dataclass
class ExceptReturnRaiseRatio:
    """Except-block return/raise ratio primitive.

    Implements ``ShapePrimitive[_ExceptRatioBaseline]``.
    """

    name: str = "except_return_raise_ratio"
    min_cluster_size: int = 10
    cluster_bonus_clip: float = 5.0

    def fit_cluster_baseline(
        self,
        cluster_files: Iterable[tuple[Path, str]],
        language: Language,
    ) -> _ExceptRatioBaseline | None:
        """Compute mean ± std of per-file except-block ratios.

        Returns ``None`` if fewer than ``_MIN_VALID_FILES`` files have at
        least one return/raise inside a handler block (insufficient sample).
        """
        ratios: list[float] = []
        for _path, source in cluster_files:
            r = _ratio_for_source(source, language)
            if r is not None:
                ratios.append(r)
        if len(ratios) < _MIN_VALID_FILES:
            return None
        mean = statistics.mean(ratios)
        std = statistics.pstdev(ratios) if len(ratios) > 1 else 0.0
        return _ExceptRatioBaseline(mean=mean, std=std)

    def score(
        self,
        hunk_content: str,
        *,
        baseline: _ExceptRatioBaseline | None,
        cluster_size: int,
    ) -> float:
        """Return tail-z contribution clipped to ``cluster_bonus_clip``.

        Abstains (returns 0.0) when:
        - ``baseline`` is None (cluster had insufficient sample),
        - ``cluster_size < self.min_cluster_size`` (size floor),
        - hunk has no return/raise inside handler blocks (math undefined).

        Ramp: ``min(clip, max(0.0, |tail_z| - 1.0))``.
        """
        if baseline is None or cluster_size < self.min_cluster_size:
            return 0.0
        hunk_ratio = _ratio_for_hunk(hunk_content)
        if hunk_ratio is None:
            return 0.0
        tail_z = (hunk_ratio - baseline.mean) / max(baseline.std, 1e-6)
        return min(self.cluster_bonus_clip, max(0.0, abs(tail_z) - 1.0))


__all__ = ["ExceptReturnRaiseRatio"]

File: crates/argot-core/src/scoring/shape_primitives/except_return_raise_ratio.rs
//! Except-block return/raise ratio — port of
//! `engine/argot/scoring/scorers/except_return_raise_ratio.py`.
//!
//! Within each handler subtree (`except_clause` / `catch_clause`),
//! ratio = returns / (returns + raises). Nested handlers are counted
//! independently (each handler's full subtree is walked). Per-file mean of
//! handler ratios; two-sided tail-z ramp.

use crate::scoring::adapters::Language;
use crate::scoring::shape_primitive::{Baseline, ShapePrimitive};
use crate::scoring::shape_primitives::{parse, population_mean_std};
use std::path::PathBuf;
use tree_sitter::Node;

const MIN_VALID_FILES: usize = 3;

/// The exception-handler clause node kind for `language`, or `None` for
/// languages with no exception-handling construct (Go, Rust, C — errors are
/// values there, so there are no handler blocks to measure).
fn handler_kind(language: Language) -> Option<&'static str> {
    match language {
        Language::Python => Some("except_clause"),
        Language::Typescript
        | Language::Javascript
        | Language::Java
        | Language::CSharp
        | Language::Cpp
        | Language::Php => Some("catch_clause"),
        Language::Ruby => Some("rescue"),
        Language::Go | Language::Rust | Language::C => None,
    }
}

/// The `return`-statement node kind for `language`.
fn return_kind(language: Language) -> &'static str {
    match language {
        Language::Ruby => "return",
        _ => "return_statement",
    }
}

/// Whether `node` raises. Python/TS have a dedicated statement node; Ruby
/// raises via a `raise`/`fail` method call, so it needs the source text.
fn is_raise(node: Node, src: &[u8], language: Language) -> bool {
    match language {
        Language::Python => node.kind() == "raise_statement",
        // PHP raises via a throw *expression*, not a statement.
        Language::Php => node.kind() == "throw_expression",
        Language::Ruby => {
            node.kind() == "call"
                && node.child_by_field_name("receiver").is_none()
                && node
                    .child_by_field_name("method")
                    .map(|m| {
                        let text = &src[m.byte_range()];
                        text == b"raise" || text == b"fail"
                    })
                    .unwrap_or(false)
        }
        // TypeScript, Java, C#, C++ all raise via a throw statement. Go, Rust
        // and C never reach here (they have no handler kind).
        _ => node.kind() == "throw_statement",
    }
}

/// Count `(returns, raises)` in the whole subtree rooted at `root`.
fn count_returns_raises(root: Node, src: &[u8], language: Language) -> (usize, usize) {
    let return_type = return_kind(language);
    let mut returns = 0usize;
    let mut raises = 0usize;
    let mut stack = vec![root];
    while let Some(node) = stack.pop() {
        if node.kind() == return_type {
            returns += 1;
        } else if is_raise(node, src, language) {
            raises += 1;
        }
        let n = node.child_count();
        for i in (0..n).rev() {
            if let Some(c) = node.child(i) {
                stack.push(c);
            }
        }
    }
    (returns, raises)
}

/// Sum `(returns, raises)` across every handler subtree under `root`. Nested
/// handlers are walked independently (double-counting is intentional, matching
/// the Python `_count_in_handlers`).
fn count_in_handlers(root: Node, src: &[u8], language: Language) -> (usize, usize) {
    let handler_type = match handler_kind(language) {
        Some(h) => h,
        None => return (0, 0),
    };
    let mut returns = 0usize;
    let mut raises = 0usize;
    let mut stack = vec![root];
    while let Some(node) = stack.pop() {
        if node.kind() == handler_type {
            let (r, x) = count_returns_raises(node, src, language);
            returns += r;
            raises += x;
        }
        let n = node.child_count();
        for i in (0..n).rev() {
            if let Some(c) = node.child(i) {
                stack.push(c);
            }
        }
    }
    (returns, raises)
}

fn ratio_for_source(source: &str, language: Language) -> Option<f64> {
    let tree = parse(source, language)?;
    // Languages with no exception-handling construct (Go, Rust, C) have no
    // handler blocks, so the ratio is undefined — no signal.
    handler_kind(language)?;
    let (returns, raises) = count_in_handlers(tree.root_node(), source.as_bytes(), language);
    let total = returns + raises;
    if total == 0 {
        return None;
    }
    Some(returns as f64 / total as f64)
}

/// Probe the exception-having grammars in turn; first defined ratio wins.
/// The hunk's language is not known at score time, so we try each grammar
/// that has an exception-handler construct.
fn ratio_for_hunk(hunk: &str) -> Option<f64> {
    ratio_for_source(hunk, Language::Python)
        .or_else(|| ratio_for_source(hunk, Language::Typescript))
        .or_else(|| ratio_for_source(hunk, Language::Java))
        .or_else(|| ratio_for_source(hunk, Language::CSharp))
        .or_else(|| ratio_for_source(hunk, Language::Php))
        .or_else(|| ratio_for_source(hunk, Language::Cpp))
        .or_else(|| ratio_for_source(hunk, Language::Ruby))
}

/// Except-block return/raise ratio primitive.
#[derive(Default)]
pub struct ExceptReturnRaiseRatio;

impl ShapePrimitive for ExceptReturnRaiseRatio {
    fn name(&self) -> &str {
        "except_return_raise_ratio"
    }
    fn min_cluster_size(&self) -> usize {
        10
    }
    fn cluster_bonus_clip(&self) -> f64 {
        5.0
    }

    fn fit_cluster_baseline(
        &self,
        cluster_files: &[(PathBuf, String)],
        language: Language,
    ) -> Option<Baseline> {
        let mut ratios: Vec<f64> = Vec::new();
        for (_path, source) in cluster_files {
            if let Some(r) = ratio_for_source(source, language) {
                ratios.push(r);
            }
        }
        if ratios.len() < MIN_VALID_FILES {
            return None;
        }
        let (mean, std) = population_mean_std(&ratios);
        Some(Baseline::MeanStd { mean, std })
    }

    fn score(&self, hunk: &str, baseline: Option<&Baseline>, cluster_size: usize) -> f64 {
        let Some(Baseline::MeanStd { mean, std }) = baseline else {
            return 0.0;
        };
        if cluster_size < self.min_cluster_size() {
            return 0.0;
        }
        let Some(hunk_ratio) = ratio_for_hunk(hunk) else {
            return 0.0;
        };
        let tail_z = (hunk_ratio - *mean) / std.max(1e-6);
        (tail_z.abs() - 1.0).max(0.0).min(self.cluster_bonus_clip())
    }
}
